[CIR] Call code gen; create empty cir.func op

dkolsen-pgi opened the pull request:
Finish hooking up ClangIR code gen into the Clang control flow, initializing enough that basic code gen is possible.

Add an almost empty `cir.func` op to the ClangIR dialect.  Currently the only property of the function is its name.  Add the code necessary to code gen a cir.func op.

Create essentially empty files
clang/lib/CIR/Dialect/IR/{CIRAttrs.cpp,CIRTypes.cpp}.  These will be filled in later as attributes and types are defined in the ClangIR dialect.

(Part of upstreaming the ClangIR incubator project into LLVM.)

dkolsen-pgi committed:
[CIR] Call code gen; create empty cir.func op

Finish hooking up ClangIR code gen into the Clang control flow,
initializing enough that basic code gen is possible.

Add an almost empty cir.func op to the ClangIR dialect.  Currently the
only property of the function is its name.  Add the code necessary to
code gen a cir.func op.

Create essentially empty files
clang/lib/CIR/Dialect/IR/{CIRAttrs.cpp,CIRTypes.cpp}.  These will be
filled in later as attributes and types are defined in the ClangIR
dialect.

(Part of upstreaming the ClangIR incubator project into LLVM.)

SchrodingerZhu commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -16,4 +16,50 @@
+// CIR Ops
+//===----------------------------------------------------------------------===//
+
+class LLVMLoweringInfo {
```
Document for this was removed?

dkolsen-pgi replied:
I was intended to migrate the documentation for `LLVMLoweringInfo` later.  But looking it over, there is no good reason to not upstream it now.  I have fixed this in my workspace (adding the long comment) and will commit later with other changes still pending for this PR.

SchrodingerZhu commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -16,4 +16,50 @@
+// For starters, cir.func has only name, nothing else.  The other properties
+// of a function will be added over time as more of ClangIR is upstreamed.
+
+def FuncOp : CIR_Op<"func"> {
```
Maybe mention that some traits are to be added

dkolsen-pgi replied:
That's exactly what the comment says, "The other properties of a function will be added over time."

I'm still trying to find the right balance for comments explaining what has not yet been implemented (or upstreamed in this case).  Too many such comments is just clutter and gets in the way.  Too few is not helpful.

smeenai commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -24,9 +27,149 @@ CIRGenModule::CIRGenModule(mlir::MLIRContext &context,
+}
+
+void CIRGenModule::buildGlobal(clang::GlobalDecl gd) {
+
```
Nit: extra newline

smeenai commented on clang/lib/CIR/FrontendAction/CIRGenAction.cpp:
```diff
@@ -37,14 +40,39 @@ class CIRGenConsumer : public clang::ASTConsumer {
 
+  void Initialize(ASTContext &Ctx) override {
+    assert(!Context && "initialized multiple times");
+
+    Context = &Ctx;
+
```
The newlines seem superfluous, unless there's more to be filled in here later.

smeenai reviewed (commented):
Very minor nits :)

Is it at all possible to write a test to exercise the new functionality, or will it just run into an NYI?

dkolsen-pgi commented:
> Is it at all possible to write a test to exercise the new functionality,

Yes, I am planning to write a CIR code gen test as part of this PR, to verify that a `cir.func` op is created.  But there is some broken or missing functionality that I need to fix first before I can even write the test.

keryell commented on clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:
```diff
@@ -0,0 +1,38 @@
+
+#include "clang/CIR/Dialect/IR/CIRDialect.h"
+
+using namespace mlir;
+using namespace mlir::cir;
```
What about removing all the `using namespace` since this is a new start to the project?

dkolsen-pgi replied:
Those `using namespace` directives are in the Clang incubator project.  I would prefer to minimize the gratuitous differences between LLVM upstream and ClangIR incubator.

There is discussion about getting rid of the `::mlir::cir` namespace and moving everything therein to the `::cir` namespace.  If that happens, I will make the corresponding change upstream and will likely get rid of the `using namespace` directives then.

keryell reviewed (commented):
Nice!

dkolsen-pgi committed:
[CIR] Call code gen; create empty cir.func op

Followup commit:

Improve comments, variable names, and formatting.  Some other code
cleanup.

Fix bugs in the creation of the `cir.func` op: add it to the module, and
correctly store its name.

Update the hello.c test to check that a `cir.func` op is generated.

dkolsen-pgi commented:
> Is it at all possible to write a test to exercise the new functionality,

Yes, The change that I just committed includes updating the existing test to verify that a `cir.func` op is generated correctly.

dkolsen-pgi committed:
Merge branch 'main' into cir/upstream/func

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -24,9 +27,140 @@ CIRGenModule::CIRGenModule(mlir::MLIRContext &context,
+void CIRGenModule::buildGlobalFunctionDefinition(clang::GlobalDecl gd,
+                                                 mlir::Operation *op) {
+  auto const *funcDecl = cast<FunctionDecl>(gd.getDecl());
+  auto funcOp = builder.create<mlir::cir::FuncOp>(
```
This is called `buildGlobalFunctionDefinition`, but never looks a the definition?  Is this intended to be just for 'defined' functions, or do we mean declaration?

lanza replied:
The names of much of the functions in CIRGen are lifted directly from CodeGen. This fn isn't fleshed out yet so it's not going to fully honor the name here, but it's structurally what should happen later. 

See here: https://github.com/llvm/clangir/blob/main/clang/lib/CIR/CodeGen/CIRGenModule.cpp#L601

dkolsen-pgi replied:
The code to process the function definition exists in the ClangIR incubator and will be upstreamed later.  With this change, `cir.func` op can't represent function bodies, and there is not CIR code gen code to process the AST for function bodies.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -24,9 +27,140 @@ CIRGenModule::CIRGenModule(mlir::MLIRContext &context,
+
+DiagnosticBuilder CIRGenModule::errorNYI(llvm::StringRef feature) {
+  unsigned diagID = diags.getCustomDiagID(DiagnosticsEngine::Error,
+                                          "ClangIR code gen NYI: %0");
```
```suggestion
                                          "ClangIR code gen Not Yet Implemented: %0");
```

Same for the rest here, I'd prefer we are better about this for user/early adopter readability.

dkolsen-pgi replied:
Done.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -24,9 +27,140 @@ CIRGenModule::CIRGenModule(mlir::MLIRContext &context,
+
+DiagnosticBuilder CIRGenModule::errorNYI(SourceRange loc,
+                                         llvm::StringRef feature) {
+  unsigned diagID = diags.getCustomDiagID(DiagnosticsEngine::Error,
```
These should probably just call the version on 138/145.

dkolsen-pgi replied:
Done.

erichkeane commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -10,4 +10,57 @@
+
+#include "mlir/Support/LogicalResult.h"
+
+#include "clang/CIR/Dialect/IR/CIROpsDialect.cpp.inc"
```
Alphabetization concerns again.

dkolsen-pgi replied:
The first of the three includes here is the header file for this source file.  In some coding standards that header should always be first.  Does Clang have that convention?

The last of the three includes is the TableGen-generated implementation file.  I feel like that should always be last.  Is there a Clang or LLVM convention about where that should go?

That leaves only one include, the middle one, that needs to be alphabetized.

erichkeane replied:
Here's our policy https://llvm.org/docs/CodingStandards.html#include-style

So tablegen files don't seem to be special cases? Current header seems like it should be first.

dkolsen-pgi replied:
The TableGen'erated file doesn't have its own includes.  So it needs to be last, to rely on the includes of the main `.cpp` file.  The TableGen file is not a regular header and shouldn't be lumped in with the guidelines for the other headers.

AaronBallman replied:
> The TableGen'erated file doesn't have its own includes. So it needs to be last, to rely on the includes of the main .cpp file.

I think that should probably change (though I don't think it needs to as part of this patch) because relying on transitive includes is rarely a good thing for long-term maintenance.

erichkeane reviewed (commented):
A few nits along the way.  Thanks for the patch!

lanza commented on clang/include/clang/CIR/CIRGenerator.h:
```diff
@@ -53,6 +53,7 @@ class CIRGenerator : public clang::ASTConsumer {
   ~CIRGenerator() override;
   void Initialize(clang::ASTContext &astCtx) override;
   bool HandleTopLevelDecl(clang::DeclGroupRef group) override;
+  mlir::ModuleOp getModule();
```
For Code/CIR-Gen there isn't much reference to anything other than the MLIR or LLVM module. I think this is fine. However, for the contexts we do need to change them to be specific. Whenever we upstream them it should specifically say `get{MLIR,AST}Context`.

erichkeane replied:
Ok, that seems fair.  We definitely have a few places where `getContext` means something ambiguous, so wanted to make sure this WOULDN'T be.

lanza commented on clang/include/clang/CIR/Dialect/IR/CIRDialect.h:
```diff
@@ -13,4 +13,22 @@
+
+#include "clang/CIR/Dialect/IR/CIROpsDialect.h.inc"
+
+#define GET_OP_CLASSES
```
That's just an MLIR mechanic, not custom to CIR. But I agree that it probably is reasonable to have a comment describing it.

dkolsen-pgi replied:
I added a comment.

AaronBallman replied:
Thank you for that comment, I was quite confused as well!

lanza commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -24,9 +27,140 @@ CIRGenModule::CIRGenModule(mlir::MLIRContext &context,
                            clang::ASTContext &astctx,
```
The .cir-tidy configuration file has been added to each directory in ClangIR so clangd should yell at you if it's wrong. But this is in agreement with the previous conversation -- CIRGen is to be stylized like mlir and not clang.

erichkeane replied:
Ok, thanks!  Wanted to make sure this was discussed.  I'll try to remember this at least until the next patch :)

lanza commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -24,9 +27,140 @@ CIRGenModule::CIRGenModule(mlir::MLIRContext &context,
+    if (fd->hasAttr<AnnotateAttr>()) {
+      errorNYI(fd->getSourceRange(), "defferedAnnotations");
+    }
+    if (!fd->doesThisDeclarationHaveABody()) {
```
This is in deciding what to do with specifically the current decl being handed us over the parseTopLevelDecl -> handleTopLevelDecl walk. Later in the callstack specific things are done to make sure we're looking at the right thing. And all the deferment systems from CodeGen exists for CIR, too.

dkolsen-pgi commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -24,9 +27,140 @@ CIRGenModule::CIRGenModule(mlir::MLIRContext &context,
+  StringRef filename = pLoc.getFilename();
+  return mlir::FileLineColLoc::get(builder.getStringAttr(filename),
+                                   pLoc.getLine(), pLoc.getColumn());
+}
```
I am guessing that the MLIR location type doesn't have any notion of multiple nested locations due to macro expansions.  So we can't represent the full complexity of macros here.  The outermost location seems like the right one to choose.  (I am not familiar with either the Clang location type or the MLIR location type, so I welcome being corrected here.)

erichkeane replied:
I'll leave this open for Aaron to comment on.  IIRC, the source location stored in the AST is going to be the location of the macro, which can make diagnostics awkward/confusing if what happens is inside the expansion, but if MLIR can't comprehend "in expansion of..." then this is perhaps as good as we can do.

AaronBallman replied:
I think this may be okay -- a `PresumedLoc` is sensitive to `#line` and GNU line markers, etc, and it's always the expansion point of the wrapped source location. However, MLIR may eventually need to grow to understand the difference between spelling, expansion, and presumed locations.

efriedma-quic replied:
If you ever want to generate diagnostics from CIR, you probably want to just use the clang SourceLocation directly, not try to translate it to an MLIR "Location".  It's extremely complicated to describe a location when macros are involved, and the abstraction clang uses really only makes sense for C preprocessor.

dkolsen-pgi replied:
Diagnostics that happen during CIR code generation will use the Clang `SourceLocation` because the Clang AST is still available.  But later passes won't have access to a `SourceLocation`.  Every MLIR node already has an `mlir::Location`.  Storing a `SourceLocation` instead of, or in addition to, an `mlir::Location` is not practical.

I don't expect many diagnostics to be issued from CIR when compilation.  Virtually all diagnostics of interest should have been issued during parsing and semantic checking.  The answer may be different when CIR is used for static analysis.  I'll let the static analysis experts sort that out.

dkolsen-pgi replied:
> I don't buy the reasoning of "mlir::Location exists, therefore it's the right representation".

`mlir::Location` is not the right representation because it exists.  It's because it is used by all the other MLIR dialects that have nothing to do with C++ and cannot have a dependency on Clang.  It is wanting ClangIR to play nicely with the rest of the MLIR ecosystem, when many dialects are mixed together in a single tree later in the compilation chain, that makes using `SourceLocation` problematic.

> Not saying this is something we need to solve right now

Definitely.  Using `SourceLocation` instead of `mlir::Location` is not something that can be done in this PR.  We would need evidence that the lack of `SourceLocation` and the richness of what it can represent is a serious problem before attempting such a radical change.

AaronBallman replied:
>  It is wanting ClangIR to play nicely with the rest of the MLIR ecosystem, when many dialects are mixed together in a single tree later in the compilation chain, that makes using SourceLocation problematic.

However, that's also an internal implementation detail that doesn't need to be exposed to users of the CIR library. Clang works in `SourceLocation`; if CIR needs to work in something else, that should be abstracted away from the user. Still not saying this needs to be solved as part of this patch, but `CIRGenModule::getLoc()` are public interfaces currently; can those be made private so we're not leaking an implementation detail of CIR to users?

> But later passes won't have access to a SourceLocation.

Two things:

1) Diagnostics that depend on optimization levels have massive usability issues, so diagnostics from later passes strikes me as a potential design concern to start with; hopefully we don't end up in that situation except in very rare circumstances.
2) The backend does occasionally need to issue user-facing diagnostics (things like `__attribute__((error))` come to mind) and this has been a huge source of pain because LLVM does a pretty bad job of tracking source location with sufficient fidelity for Clang to have good QoI. If MLIR has the same issues, that should be something the MLIR folks work to correct before we transition to CIR as a default.

dkolsen-pgi replied:
> However, that's also an internal implementation detail that doesn't need to be exposed to users of the CIR library. 

The primary use of ClangIR is intended to be of the MLIR data structures after CIR code gen has happened.  The public interface for ClangIR is entirely based on MLIR, including `mlir::Location`.  Some users will want to use the CIR library to convert from the Clang AST to the ClangIR MLIR dialect, but for many users of ClangIR the fact that the MLIR data structures were generated from the Clang front end and the Clang AST is an implementation detail that they don't need to know about.  `SourceLocation` is more of an implementation detail than `mlir::Location` is.

Changing `CIRGenModule::getLoc()` to be private functions would be problematic because those functions will be used by many different classes within the ClangIR code gen code.  Those functions really should have module accessibility (similar to package or assembly accessibility in Java or C#), but that's not practical until Clang/LLVM fully moves to C++20 modules.

I am aware and agree that (1) back end diagnostics should be avoided where possible and should be rare, but that (2) when they do happen they can be hard to get right.  We struggle with that same problem with NVC++.  We should solve the problem of getting good diagnostics during the MLIR passes when we actually run into the problem.  It is only then that we will know how serious the problem is, how best to solve it, and how much effort to put into it.  Solving that problem now is premature and will miss the mark in some way.

AaronBallman replied:
>  SourceLocation is more of an implementation detail than mlir::Location is.

That depends on perspective, IMO. From our perspective, CIR is the way in which we lower Clang's AST to an IR that eventually ends up in LLVM. The public entrypoints into CIR from Clang's perspective should be using Clang's data structures. If those need to be converted internally into MLIR-specific data structures, that's fine, but that should not leak out into the public interfaces that Clang can interact with.

However, I see now that I missed something important here -- this code is under `clang/lib/CIR` and not `clang/include/clang/CIR`, so it *isn't* in the public interfaces that Clang interacts with, it is (sufficiently) hidden as an implementation detail. Sorry for missing that!

> We should solve the problem of getting good diagnostics during the MLIR passes when we actually run into the problem. It is only then that we will know how serious the problem is, how best to solve it, and how much effort to put into it. Solving that problem now is premature and will miss the mark in some way.

The point to these initial PRs is to ensure the community is comfortable with the design and doesn't see any glaring design concerns, so saying "we'll solve it when we get to it" on a situation we know we will hit sort of defeats the purpose. We already know the problem exists because it's one we fight with today and we already know a better way to solve it is to use source location information with at least as much fidelity as Clang is able to handle. Given that one of the benefits to CIR is for better codegen analysis that can lead to improved diagnostics, having this as an early design concern is pretty reasonable IMO. That said, we're not insisting on a change as part of this PR -- just that the CIR folks understand that this is a design concern with the facilities and at some point it may go from "concern" to "required to improve before we can proceed."

SchrodingerZhu commented:
Does https://mlir.llvm.org/docs/Dialects/Builtin/#fusedloc
suit the situation? 
On Thu, Oct 31, 2024 at 15:48, David Olsen ***@***.***&gt; wrote:   @dkolsen-pgi commented on this pull request.  In clang/lib/CIR/CodeGen/CIRGenModule.cpp: &gt; @@ -24,9 +27,140 @@ CIRGenModule::CIRGenModule(mlir::MLIRContext &amp;context,
                            clang::ASTContext &amp;astctx,
                            const clang::CodeGenOptions &amp;cgo,
                            DiagnosticsEngine &amp;diags)
-    : astCtx(astctx), langOpts(astctx.getLangOpts()),
-      theModule{mlir::ModuleOp::create(mlir::UnknownLoc())},
-      target(astCtx.getTargetInfo()) {}
+    : builder(&amp;context), astCtx(astctx), langOpts(astctx.getLangOpts()),
+      theModule{mlir::ModuleOp::create(mlir::UnknownLoc::get(&amp;context))},
+      diags(diags), target(astCtx.getTargetInfo()) {}
+
+mlir::Location CIRGenModule::getLoc(SourceLocation cLoc) {
+  assert(cLoc.isValid() &amp;&amp; "expected valid source location");
+  const SourceManager &amp;sm = astCtx.getSourceManager();
+  PresumedLoc pLoc = sm.getPresumedLoc(cLoc);
+  StringRef filename = pLoc.getFilename();
+  return mlir::FileLineColLoc::get(builder.getStringAttr(filename),
+                                   pLoc.getLine(), pLoc.getColumn());
+}
 I am guessing that the MLIR location type doesn't have any notion of multiple nested locations due to macro expansions. So we can't represent the full complexity of macros here. The outermost location seems like the right one to choose. (I am not familiar with either the Clang location type or the MLIR location type, so I welcome being corrected here.) —
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you commented.Message ID: ***@***.***&gt;

dkolsen-pgi commented:
> Does https://mlir.llvm.org/docs/Dialects/Builtin/#fusedloc suit the situation [of recording locations of macro expansions]?

Maybe?  Right now I think `FusedLoc` is used by ClangIR to represent a source range.  Maybe it also could be used for macro expansion locations.  I don't if there is a `FusedLoc` metadata convention for that.

That would be something we want to explore and implement in the ClangIR incubator project first.  It's not something I want to implement as part of an upstreaming change.

erichkeane commented:
> > Does https://mlir.llvm.org/docs/Dialects/Builtin/#fusedloc suit the situation [of recording locations of macro expansions]?
> 
> Maybe? Right now I think `FusedLoc` is used by ClangIR to represent a source range. Maybe it also could be used for macro expansion locations. I don't if there is a `FusedLoc` metadata convention for that.
> 
> That would be something we want to explore and implement in the ClangIR incubator project first. It's not something I want to implement as part of an upstreaming change.

Understood, there is likely value to a 'FIXME' that has some sort of sub-locations/expansion locations suggested as a point of something we should fix in the future.  BUT, the current 'this is an array of locations' likely doesn't comprehend that well.

Currently when clang does diagnostics on a macro expansion, we do a bunch of notes of "in expansion of macro...", which includes the name and location of the macro being expanded.  It seems that `FusedLoc` would need some way of referring to a separate entity to get this, so I don't see it being useful for this.

Again though, a `FIXME: at one point we should capture macro expansions here so we can diagnose expansions more thoroughly` is perhaps not a bad addition here.

dkolsen-pgi committed:
[CIR] Call code gen; create empty cir.func op

Followup change responding to code review comments.  There are no
behavior changes, just small cleanup.

dkolsen-pgi commented on clang/include/clang/CIR/Dialect/IR/CIRDialect.h:
```diff
@@ -13,4 +13,22 @@
 #ifndef LLVM_CLANG_CIR_DIALECT_IR_CIRDIALECT_H
 #define LLVM_CLANG_CIR_DIALECT_IR_CIRDIALECT_H
 
+#include "mlir/IR/Builders.h"
```
`git clang-format` will alphabetize a block of includes that are not separated by blank lines.  `clang-format` puts all capital letters before all lower-case letters, so these are already in the correct order.

AaronBallman replied:
I think we'd usually have the clang includes first though so there's less chance of accidental reliance on transitive includes from mlir.

dkolsen-pgi commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -16,4 +16,87 @@
+include "mlir/Interfaces/MemorySlotInterfaces.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+include "mlir/IR/BuiltinAttributeInterfaces.td"
```
`clang-format` doesn't automatically alphabetize 'include' directives in '.td' files, so I manually changed these to be in the same order that `clang-format` would have put them in.

AaronBallman commented on clang/include/clang/CIR/CIRGenerator.h:
```diff
@@ -53,6 +53,7 @@ class CIRGenerator : public clang::ASTConsumer {
   ~CIRGenerator() override;
   void Initialize(clang::ASTContext &astCtx) override;
   bool HandleTopLevelDecl(clang::DeclGroupRef group) override;
+  mlir::ModuleOp getModule();
```
Is there a reason this isn't a `const` member function?

dkolsen-pgi replied:
https://mlir.llvm.org/docs/Rationale/UsageOfConst/

Types that implement MLIR data structures are not `const` correct.  That attitude will probably bleed over into ClangIR types that aren't part of the MLIR dialect.  But in this case I think this particular function can be `const`.

dkolsen-pgi replied:
Fixed.  I changed `CIRGenerator::getModule()` to be a `const` member function.

AaronBallman commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -24,9 +27,135 @@ CIRGenModule::CIRGenModule(mlir::MLIRContext &context,
+    // Update deferred annotations with the latest declaration if the function
+    // was already used or defined.
+    if (fd->hasAttr<AnnotateAttr>()) {
+      errorNYI(fd->getSourceRange(), "defferedAnnotations");
```
Typo: `deferredAnnotations`

Also, the curlies should be removed per coding style (and local convention).

dkolsen-pgi replied:
> the curlies should be removed per coding style

Ugh.  One constant in my 30+ years of programming is to always use `{ }` for sub-statements.  Code without the curlies just looks wrong and takes more effort to read.  It's hard for me to not fix code that doesn't use `{ }`.  It will be even harder for me to not use `{ }` in code that I write.

erichkeane replied:
>Ugh. One constant in my 30+ years of programming is to always use { } for sub-statements. Code without the curlies just looks wrong and takes more effort to read. It's hard for me to not fix code that doesn't use { }. It will be even harder for me to not use { } in code that I write.

This was definitely one of my  harder parts of the coding standard to get used to :)  I managed to relax the rule a bit at least, but this is effectively the canonical example of 'skip curleys'.

dkolsen-pgi replied:
Fixed the typo and removed the curly braces.

AaronBallman commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -24,9 +27,135 @@ CIRGenModule::CIRGenModule(mlir::MLIRContext &context,
+                                                 mlir::Operation *op) {
+  auto const *funcDecl = cast<FunctionDecl>(gd.getDecl());
+  auto funcOp = builder.create<mlir::cir::FuncOp>(
+      getLoc(funcDecl->getSourceRange()), funcDecl->getIdentifier()->getName());
```
Is this using the correct name in the presence of things which rename the function? e.g., https://godbolt.org/z/E6dr4vznj

erichkeane replied:
No, I don't think it is.  Though, I know there was discussions about `alias` offline, so this is almost definitely part of the "something we'll get to in the future" patch (as this is effectively just the 'bare minimum to show we can do a thing' patch).

AaronBallman replied:
Leaving that for follow-up work makes sense to me, thanks!

For these sort of questions, I think it might make sense to stub in a test case with a FIXME comment and doing the bare minimum to ensure the test doesn't crash. Then we don't accidentally forget to cover those cases again in the future when we're not looking at what was discussed in much earlier reviews. WDYT?

erichkeane replied:
There is value to that in the general case, but because of the way this is being upstreamed (and checked' against existing tests?), our existing tests + the downstream should be sufficient to make sure it isn't forgotten.

As far as 'test to make sure we don't crash'... I don't have a problem with adding tests, but I think the maturity of this project (as upstreamed), crashing is fairly well expected :) .

AaronBallman replied:
I don't expect the incubator project to have tested all the weird scenarios we come up with during PR review, and I don't expect reviewers to remember test cases from earlier reviews once they become more relevant, AND I don't expect CIR folks to handle every test case up front because that likely involves doing/undoing a bunch of work and makes upstreaming harder.

How about this for a compromise: as we think of test cases during PR review, the patch author either verifies the incubator project has test coverage for that already or adds the test coverage to the incubator project, and then fixes issues wherever it makes the most sense (either as part of the PR, a separate upstream PR, or in the incubator side which then gets upstreamed in a future PR)?

erichkeane replied:
That seems completely sensible.  I might add as far as the 'fixes the issue' is "files an issue on the incubator"?  But otherwise agree.

dkolsen-pgi replied:
This is placeholder code, as Erich said the bare minimum to not choke on function definitions.  There are large amounts of code in the ClangIR incubator that will be upstreamed later to handle function definitions and their names.  Ignoring the alias issue for the moment, the name of the `cir.func` op should be the mangled name of the function, not the simple name.  But the necessary ABI classes that do the mangling haven't been upstreamed yet.

If there are existing Clang tests for edge cases (such as the `alias` attribute), then the feature won't be forgotten during ClangIR development.  Eventually every Clang test will have to be handled correctly by the ClangIR compilation path, so missing features will be noticed.  I strongly object to expecting me to create a stub test in the LLVM project; that would be a lot of effort for no real benefit.  I appreciate that that is no longer the request.  I would prefer to not worry about things like this at all.  But a compromise I would be fine with is for the PR owner to check that either (1) the feature is implemented and tested in the ClangIR incubator project, or (2) there is an issue to implement the feature in the ClangIR incubator project.

erichkeane replied:
> (1) the feature is implemented and tested in the ClangIR incubator project, or (2) there is an issue to implement the feature in the ClangIR incubator project.

I believe this is consistent with what I suggested.  We can't really depend on our existing tests, as they are... questionable in some cases, so I'd like to make sure it gets captured on the CIR project. IF it is actually already implemented though, I expect that the PR owner should write any missing test coverage.

erichkeane replied:
In Summary, so that we all can agree here, this is the chart of what I'm suggesting:

- Questioned example "Works" in the llvm/llvm-project + this patch:
  - Test coverage exists: *PR Owner* points this out
  - Test coverage does NOT exist; *PR Owner* adds requested test
- Doesn't work in llvm/llvm-project, but works in CIR Incubator:
  - Test coverage exists: *PR Owner* points this out
  - Test coverage does NOT exist: *PR owner* adds the test to the incubator.
- Doesn't work in either
  -  The *PR Owner* Files an issue against the Incubator

dkolsen-pgi replied:
That should work, with one slight change.
- Doesn't work in either
  - The _PR Owner_ points out an existing issue in the incubator or files an issue against the Incubator.

dkolsen-pgi replied:
The ClangIR incubator currently does not support the `alias` attribute as used in Aaron's example at the beginning of this thread.  I created https://github.com/llvm/clangir/issues/1048 so that feature is not forgotten.  I think that's enough to resolve this particular comment.

AaronBallman commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -24,9 +27,135 @@ CIRGenModule::CIRGenModule(mlir::MLIRContext &context,
+void CIRGenModule::buildTopLevelDecl(Decl *decl) {
+
+  // Ignore dependent declarations.
+  if (decl->isTemplated())
```
@erichkeane -- should this be looking at the type to determine if it's dependent rather than looking at the declaration? e.g., `void foo(auto a);` is that going to return `false` for `isTemplated()`?

erichkeane replied:
We definitely want to be looking at the `Decl`, a function can be a template that has a concrete type.

```
template<typename T>
void foo() {
}
```
has a non-dependent type `void()()`, but is a template still.  This is a simple example, but things like non-template member functions of a class template do the same thing.

Also, I checked, and this is basically the first  3 lines of `CodeGenModule::EmitTopLevelDecl` (plus a `\n`).

AaronBallman replied:
Excellent, thank you!

AaronBallman commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -24,9 +27,135 @@ CIRGenModule::CIRGenModule(mlir::MLIRContext &context,
+    auto *fd = cast<FunctionDecl>(decl);
+    // Consteval functions shouldn't be emitted.
+    if (!fd->isConsteval())
+      buildGlobal(fd);
```
What happens if the function is declared but never defined?

erichkeane replied:
In the future the call to `emitGlobalFunctionDefinition` would just skip it if there is no definition if this matches current CodeGen.

AaronBallman replied:
SGTM modulo the question about trivial test coverage

dkolsen-pgi replied:
Functions that are declared but never defined will be handled in `CIRGenModule::buildGlobal`.  There is a little bit of stub code for that in this PR:
```
      errorNYI(fd->getSourceRange(),
               "function declaration that forces code gen");
```
There is code in the incubator to handle this that will be upstreamed later.

AaronBallman commented on clang/lib/CIR/CodeGen/CIRGenModule.h:
```diff
@@ -52,10 +62,33 @@ class CIRGenModule : public CIRGenTypeCache {
 public:
+  mlir::ModuleOp getModule() const { return theModule; }
+
+  /// Helpers to convert Clang's SourceLocation to an MLIR Location.
```
This should probably document whether the spelling or expansion location is what's expected?

dkolsen-pgi replied:
I am not familiar with the complexities of Clang's source locations, so I don't know what to document.  I am open to suggestions about what to put here.

AaronBallman replied:
I don't know the details about CIR, so it's hard for me to say for certain. Maybe an example will help tease this out? Given:
```
#define ten 10
#define foo(x) (x + ten)
int y = foo(12);
```
you get source locations like: https://godbolt.org/z/oG7fq9b1K Notice how you get `Loc=<<source>:3:9 <Spelling=<source>:2:16>>` or similar where the source location of the final expanded token and the spelling location of where the pre-preprocessed token lives are different -- which location does MLIR expect to work with? It could work with the spelling location (where tokens physically reside in the source file), or it could work with the expansion locations (where the tokens logically reside in the source file after preprocessing), or it could be ambivalent about which form it uses and will happily work with whatever.

dkolsen-pgi replied:
I ran some tests.  ClangIR (both the incubator and this PR) uses the presumed location from the `SourceLocation` when creating an `mlir::Location`.  It honors `#line` directives and records the location where the macro was expanded.  That is the location we most want to keep around.

I changed the comment to say that the presumed location is what is converted.

bcardosolopes replied:
What we get in this case is:
```
module ...  {
  cir.global  external dsolocal @y = #cir.int<22> : !s32i {alignment = 4 : i64} loc(#loc3)
} loc(#loc)
#loc = loc("a.cpp":0:0)
#loc1 = loc("a.cpp":3:1)
#loc2 = loc("a.cpp":3:9)
#loc3 = loc(fused[#loc1, #loc2])
```

> which location does MLIR expect to work with?

Whatever we tell it should hold. We usually take the presumed (expanded) loc (from `getSourceRange` or begin/end):

```
mlir::Location CIRGenFunction::getLoc(SourceLocation SLoc) {
  // Some AST nodes might contain invalid source locations (e.g.
  // CXXDefaultArgExpr), workaround that to still get something out.
  if (SLoc.isValid()) {
    const SourceManager &SM = getContext().getSourceManager();
    PresumedLoc PLoc = SM.getPresumedLoc(SLoc);
    StringRef Filename = PLoc.getFilename();
    return mlir::FileLineColLoc::get(builder.getStringAttr(Filename),
                                     PLoc.getLine(), PLoc.getColumn());
  } else {
    // Do our best...
    assert(currSrcLoc && "expected to inherit some source location");
    return *currSrcLoc;
  }
}
```
My take is that this is good enough for CIRGen purposes and can/should probably be improved when we start working on emitting debug information (where we can go for non-presumed locs if it makes sense), before that it feels designing prematurely to me.

AaronBallman commented on clang/lib/CIR/CodeGen/CIRGenerator.cpp:
```diff
@@ -31,9 +34,14 @@ void CIRGenerator::Initialize(ASTContext &astCtx) {
 
-  cgm = std::make_unique<CIRGenModule>(*mlirCtx, astCtx, codeGenOpts, diags);
+  mlirCtx = std::make_unique<mlir::MLIRContext>();
+  mlirCtx->getOrLoadDialect<mlir::cir::CIRDialect>();
```
I'm surprised to see something that's named `get()` and doesn't actually look at the return value or accept any arguments; what's going on here?

lanza replied:
There's a `loadDialect` equivalent that we can switch to.

dkolsen-pgi replied:
Fixed.  Changed `getOrLoadDialect` to `loadDialect`.

dkolsen-pgi committed:
[CIR] Call code gen; create empty cir.func op

More followup responding to review comments.  Make
`CIRGenerator::getModule()` a `const` member function. Call
`loadDialect` instead of `getOrLoadDialect` to load MLIR dialects.  Fix
a typo and remove unnecessary curly braces.

dkolsen-pgi committed:
Better comment about location conversions

Improve the comment about converting `SourceLocation` to
`mlir::Location`, being explicit that the presumed location is
converted.

Part of #113483 [CIR] Call code gen; create empty cir.func op

bcardosolopes commented:
> Again though, a `FIXME: at one point we should capture macro expansions here so we can diagnose expansions more thoroughly` is perhaps not a bad addition here.

+1

AaronBallman reviewed (approved):
LGTM

erichkeane commented:
@lanza and @bcardosolopes are code owners here, so we need a +1 from one of you two.  I'll squash/merge for David after that happens (and when he/I are ready for it:) ).

bcardosolopes reviewed (approved):
Thanks David!
